### PR TITLE
xds: parse ClusterSpecifierPlugin from RouteConfiguration in xDS response

### PR DESCRIPTION
**Pull request by dapengzhang0**
Implement the xds Client section of [go/grpc-rls-in-xds#heading=h.9kitavdfxxiw](http://go/grpc-rls-in-xds#heading=h.9kitavdfxxiw)

### CODE REVIEW

**Comment by dapengzhang0** [comment]
Merged the fix and refactored accordingly. Now is ready for review.

File: xds/src/main/java/io/grpc/xds/ClientXdsClient.java [hunk 1]
```diff
@@ -1429,21 +1442,75 @@ public void handleRdsResponse(
         Collections.<String>emptySet(), versionInfo, nonce, errors);
   }
 
-  private static RdsUpdate processRouteConfiguration(
+  private static List<VirtualHost> processRouteConfiguration(
```
Comment (sergiitk) [inline xds/src/main/java/io/grpc/xds/ClientXdsClient.java:1445]: I'd keep the return type of `processRouteConfiguration()` to keep consistency with other resources handlers / `process*` functuions:

- `handleLdsResponse()` -> `LdsUpdate processClientSideListener()` / `LdsUpdate processServerSideListener`
- `handleCdsResponse()` -> `CdsUpdate parseCluster()` (hm, noticed  this name is out of pattern. will fix it)
- `handleRdsResponse()` -> `RdsUpdate processRouteConfiguration()`
- `handleEdsResponse()` -> `EdsUpdate processClusterLoadAssignment()`

I think effort worth for the sake of the predictability and fulfilling code reader's expectations. Even if `processRouteConfiguration` simply returns `new RdsUpdate(parseVirtualHosts(...))`

File: xds/src/main/java/io/grpc/xds/ClientXdsClient.java [hunk 2]
```diff
@@ -1429,21 +1442,75 @@ public void handleRdsResponse(
...
+    try {
+      if (typeUrl.equals(TYPE_URL_TYPED_STRUCT_UDPA)) {
+        TypedStruct typedStruct = anyConfig.unpack(TypedStruct.class);
+        typeUrl = typedStruct.getTypeUrl();
+        rawConfig = typedStruct.getValue();
+      } else if (typeUrl.equals(TYPE_URL_TYPED_STRUCT)) {
+        com.github.xds.type.v3.TypedStruct newTypedStruct =
+            anyConfig.unpack(com.github.xds.type.v3.TypedStruct.class);
+        typeUrl = newTypedStruct.getTypeUrl();
+        rawConfig = newTypedStruct.getValue();
```
Comment (sergiitk) [inline xds/src/main/java/io/grpc/xds/ClientXdsClient.java:1499]: I think this quote implies xds and udp TypedStruct types are compatible:

> In the case that the type URL is xds.type.v3.TypedStruct (or, for historical reasons, udpa.type.v1.TypedStruct), the inner type URL of TypedStruct will be utilized.  
> — https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/extension.proto#config-core-v3-typedextensionconfig

Then we can simplify this code by using unpackCompatibleType().
```java
TypedStruct typedStruct = unpackCompatibleType(anyConfig, TypedStruct.class, TYPE_URL_TYPED_STRUCT_UDPA, TYPE_URL_TYPED_STRUCT);
```

**Review by sergiitk — COMMENTED** [review]
LGTM, minor comments

File: xds/src/main/java/io/grpc/xds/ClientXdsClient.java (hunk 1 above)
Comment (sergiitk) [reply xds/src/main/java/io/grpc/xds/ClientXdsClient.java:1445]: `CdsUpdate parseCluster()` got renamed in https://github.com/grpc/grpc-java/commit/fa4b980e07400a66cfaf3693785f873031b20482#diff-6b019eb1d3518381d336168847890513a235ea4974fcb3c76a2e9ace43f1d460R828.

Comment (sergiitk) [reply xds/src/main/java/io/grpc/xds/ClientXdsClient.java:1445]: Created PR to rename parseCluster() back to processCluster(): #8802.

File: xds/src/main/java/io/grpc/xds/ClientXdsClient.java (hunk 2 above)
Comment (dapengzhang0) [reply xds/src/main/java/io/grpc/xds/ClientXdsClient.java]: Done.

File: xds/src/main/java/io/grpc/xds/ClientXdsClient.java (hunk 1 above)
Comment (dapengzhang0) [reply xds/src/main/java/io/grpc/xds/ClientXdsClient.java]: These methods have totally different sets/types of arguments though, only their return types and their callers have a pattern. I change the name and return type for the sake of reader's "predictability".

Comment (sergiitk) [reply xds/src/main/java/io/grpc/xds/ClientXdsClient.java]: Thank you!

**Review by sergiitk — APPROVED** [review]

**Pull request merged**